### ISSUE TITLE
Add appellate support to recap and juriscraper — Funded

### ISSUE THREAD

**Issue by freelawbot** [freelawproject/recap#83]
<a href="https://github.com/PlainSite"><img src="https://avatars.githubusercontent.com/u/8204476?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [PlainSite](https://github.com/PlainSite)**
_Saturday Jul 19, 2014 at 16:05 GMT_
_Originally opened as https://github.com/freelawproject/recap-extension-2.0/issues/10_

---

Right now Firefox has appellate parsing features and Chrome doesn't. It would be good to get them in the same state.

**Comment by freelawbot** [comment]
<a href="https://github.com/mlissner"><img src="https://avatars.githubusercontent.com/u/236970?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mlissner](https://github.com/mlissner)**
_Saturday Jul 19, 2014 at 18:16 GMT_

---

+1, marking for 2.0 release.

**Comment by mlissner** [comment]
This old ticket needs resolution. In a few weeks neither Chrome nor Firefox will support appellate. That's just the way it has to be for a bit, but the good news is it puts more pressure on getting this issue (from 2014) resolved.

**Comment by johnhawkinson** [comment]
@FiloSottile have you any time to take a look at this, since you wrote the appellate support for the now-deprecated Firefox extension?

**Comment by sjschultze** [comment]
Beginning work in https://github.com/freelawproject/recap-chrome/tree/appellate-dev

**Comment by sjschultze** [comment]
Some todos:
- add `isDocketQueryUrl` support and corresponding logic:
https://github.com/freelawproject/recap-chrome/commit/75bbdf2466f0db921e058a8ebbc53a7050cab058#diff-4e91f732f3d6c295fa99b1b972e0f53aL72

- Same for the doc1/docs1 pages:
https://github.com/freelawproject/recap-chrome/commit/75bbdf2466f0db921e058a8ebbc53a7050cab058#diff-4e91f732f3d6c295fa99b1b972e0f53aL56

**Comment by sjschultze** [comment]
Basic upload and parsing of CAVC dockets is now working in the `appellate-dev` branch, so we can do development without incurring PACER fees. Example case: https://www.courtlistener.com/docket/7625727/michael-p-long-v-robert-l-wilkie/

For reference:

- https://github.com/freelawproject/recap-chrome/commit/dc08c414cba1b4271d241766fc37e155819dadac
- https://github.com/freelawproject/juriscraper/commit/9df558fc94085db51c8baea6c15a7976d231084e
- https://github.com/freelawproject/juriscraper/commit/14923ede1c11419335e2e33e9ae66475bdb26fc9

Next step: **get PDF uploading working**

**Comment by mlissner** [comment]
Small thing, but when this launches, we need to:

 - [ ] Do a notification of some kind in the browser notifying people that appellate support is here. 

This is important because some people turn RECAP on and off as they use the courts to avoid uploading things they don't want to, and we don't want them accidentally learning that appellate is now uploading things (after they've uploaded something they didn't intend to).

**Comment by mlissner** [comment]
I forgot this, but the Court of Appeals for Veterans Claims runs a free site: 

https://efiling.uscourts.cavc.gov/cmecf/servlet/TransportRoom?servlet=CaseSearch.jsp

I don't know how different it is (if at all) from the regular appellate sites, but it might be a way to save on PACER fees.

**Comment by jgrisham** [comment]
> * [ ]  Do a notification of some kind in the browser notifying people that appellate support is here.

Would something similar be acceptable _now_, to show new/intermittent users that appellate support _does not yet exist_?


Tangentially, see my comments about documentation in the post-closeout comments on a duplicate of this issue: https://github.com/freelawproject/recap/issues/299#issuecomment-941792371

----
I'm happy to try and help _(and not just to disguise my embarrassment after spending 4 hours trying to get RECAP to upload 5th circuit documents / dockets over the past two days)_.

**Comment by mlissner** [comment]
Jim, are you a developer enough that you could evaluate what we've got? We have a version that's maybe 80% of the way there, but we haven't had the developer time to take it over the finish line. 

I *thought* we had a warning on the icon when you were at an appellate court, but I haven't checked in awhile. I didn't think we removed that though.

**Comment by jgrisham** [comment]
@mlissner: Evaluate, yes. No promises on fixes, though. ;)

I'm hardly an expert with modern tools _(especially regarding GitHub)_, but I'd be happy to dig in if someone were to point me towards any existing documentation and code branches _(and upload or forward a copy of any potentially relevant notes / e-mails)_.

_(I'd hate waste time re-inventing the wheel regarding any architectural / API research that was attempted in the past.)_

**Comment by jgrisham** [comment]
Follow-up… even if the code is only 80% complete now… perhaps a portion can be quickly activated in order to hoover up any documents that are purchased in the near-term; later we add the backend capability to automatically link those document numbers to specific docket and case information.

Alternatively, if the code for recording docket data is closer to completion than the document code, we can at least get some appellate dockets added to the database. Publicly-available dockets would still be a great public service before the actual contents of documents are available. In the meantime, if the dockets revealed document numbers, users without PACER access may be able to use that data to find selected documents hosted elsewhere on the Internet.

**Comment by mlissner** [comment]
@litewarp did some work on appellate that may be worth looking at, or perhaps not. He may chime in with details about how far along the work is. His fork is here:

https://github.com/litewarp/recap-chrome

The problem with this fork is that it has a LOT of other stuff in it, so there was no way to carefully review it before launching it to our 20k users. 

I honestly think a fresh implementation of appellate is easier than trying to tease out the appellate work in that fork (it's 400 commits behind main), but I know @litewarp put a lot of work into it, so I want to at least make sure that somebody looks at this and thinks about it while we're working on appellate.

**Comment by litewarp** [comment]
@mlissner I know I've voiced this before, but the existing recap code is so far behind modern JS that there is not going to be any simple migration path, nor a comprehensible way to add new features to recap. 

A lot of the "other" stuff in that fork (from 2+ years ago!) was an effort to move past jquery and aughts-level-monkey-patched-callbacks without changing the functionality. It's why I kept and ensured compliance with all the tests -- new formula; same great taste. 

Also, I wrote it in typescript and used a bundler. Migrations like this should *bolster* your confidence in shipping your app. Type checking produces enormous benefits as well as taking advantage of all the fanciness and improvements in modern day JavaScript. 

Idk, bugs happen on new versions.  But at this point even the "modernized" code I wrote would be somewhat stale.

I would love to continue to work on recap. But it's so frustrating to work on that I understand why you nor another developer have touched it. But that's literally the reason I wrote that fork! 

tl:dr -- I don't know how you move recap forward now without addressing that "other" stuff. Maybe time to rip of the band aid?

**Comment by mlissner** [comment]
It's not the older-style JS that was preventing work on these issues, it was the lack of time, and if I'm honest, I find dependency management, bundling, and the rest of it to be kind of a headache that _prevents_ me from working on things, because I'm always half expecting NPM to be the wrong version or some dependency to have a vulnerability or who knows what. I'm leaning towards ripping react out of CourtListener for this reason, for example, and to just use [HTMX](https://htmx.org/) instead — it's SO MUCH simpler.

But typescript is probably useful to some degree, and modern JS without the callback hell is certainly nice. Won't deny that, and I think we're slowly moving towards that.

The other thing that's coming up is [manifest v3](https://developer.chrome.com/docs/extensions/mv3/intro/), which will require a pretty big redo some time early next year. I think when we get to that we'll be wise to take another look at your branch and see how much of it we can re-use, but we've gotta do it bit by bit, however that's possible. Huge rewrites without planning are really just too scary.

**Comment by litewarp** [comment]
I wasn't saying that you need to wholly embrace the whole JS ecosystem, but it's absolutely necessary for RECAP. It's a scraping tool! It has to interact with the DOM! 

I hear you about having to manage and understand an ecosystem, but I would personally take npm (or pnpm or yarn) over pip anyday! Node / npm has never borked my system like python. I digress. The point is that you do have the capacity to learn an ecosystem, but you choose not to, whether because you're busy or have a distaste for it. I get it--I know very little about python and its package managers, but I'm also not in a business that absolutely requires I know it.

The things you mentioned (bundling, dependency management, and the rest of it) were all addressed in my branch -- in fact, the reason I did was to make things like updating security flaws easier, and adding some modern development tooling like hot reloading and advanced bundling which are just industry standard these days. It was also to entice other developers to contribute to it without having to spend a year spelunking the labyrinth of callbacks like I did, and like you're going to have to do for the v3 manifest upgrade to implement [service workers](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-overview/#service-workers). 

I recall that the only reason we did not release the modernized version as a beta two years ago was because you were (are?) adverse to non-incremental changes. I don't even think you ever bundled it yourself or tested it out at that time. Idk, you might be the only product lead for a *web-based* business in 2022 who doesn't embrace javascript. It's mind-boggling to me, but I guess you just think of RECAP as an after thought and not a place for new features? 

I mean, this PR has been open for eight(!!) years and I presented you a solution two years ago. That's also when I stopped working on improving recap -- those development improvements are massive time-savers and without them, it's not worth it to move forward on the code. Who wants to have to manually rebundle and upload the extension to chrome, then reload the page on every change? Yikes.

I get that it's a lot of tooling. Some of this could be simplified further with recent bundling improvements (esbuild, vite, etc.) and documentation. But it sounds like these improvements won't get merged until either (a) you dive in to understand it all, or (b) you're forced to do it as part of, let's say, a fire drill to meet manifest v3 upgrade deadline. Except that in the world of (b) you didn't have a year to beta test it, and you weren't under a shutdown-level deadline. Do you really think you're going to be comfortable then? 

Recap and Courtlistener are essential. That's why I'm happy to work on them. But you're not handling the nation's phone traffic or keeping planes in the air. You can afford a buggy beta or two, especially when doing so opens up big things and unblocks existing issues. So what will it take to get something off the ground here?

**Comment by mlissner** [comment]
We're making some progress on this over in #296. 

But we've got some fresh bugs in #322, #323, #324, and #325. Appellate is hard.

**Comment by litewarp** [comment]
@mlissner I will for-the-last-time chime in to say that the work you are doing now will need to be rebuilt again (and within the next six months or so!) when you upgrade to the v3 manifest. I think your hatred of npm is both warranted and a poor excuse to not modernize the recap extension.

In v3 you're going to have to [bundle external dependencies](https://developer.chrome.com/docs/extensions/mv3/mv3-migration/#remotely-hosted-code), [use fetch instead of XMLHttpRequest](https://developer.chrome.com/docs/extensions/mv3/mv3-migration/#background-service-workers), and migrate the code [to using service workers and ES modules](https://developer.chrome.com/docs/extensions/mv3/mv3-migration/#man-sw).  

It just seems short-sighted to me to do appellate recap on a separate track as the v3 migration, especially when most of my changes make v3 adoption easier.  Just trying to save you time, but 🤷

**Comment by mlissner** [comment]
I hear ya. We'll get moved over eventually, and I appreciate you pushing despite me. I'm not in a rush to get to v3 because Firefox is still working on supporting it. 

For the modernization stuff, I'm really not sure it makes a huge difference whether we do it now or later. The code we're adding right now for appellate support is all very similar to what we have for district courts, plus a bunch of helper functions that'll move across smoothly to v3 when we get to that. I don't think we're digging ourselves in deeper, in other words.

**Comment by litewarp** [comment]
You've been at this longer than me, so I trust you'll get it done. 

To my ears though, your last paragraph sounds a lot like "I'm pretty sure we're not making the boat sink any faster by adding a second floor dining room, so ...."

Gl and let me know if you need help or have questions!

**Comment by mlissner** [comment]
Ha. I think it's more like we're not making it sink faster by adding a leaf to the dining table, but we'll see. It'll get there!

**Comment by mlissner** [comment]
This is done and launched. Closing and walking away from my desk for a while.
